server/comms/link.go

```diff
@@ -92,205 +52,46 @@ type wsLink struct {
 }
 
 // newWSLink is a constructor for a new wsLink.
-func newWSLink(addr string, conn wsConnection) *wsLink {
-	return &wsLink{
-		on:           true,
-		ip:           addr,
-		conn:         conn,
-		quit:         make(chan struct{}),
-		outChan:      make(chan []byte, outBufferSize),
+func newWSLink(addr string, conn ws.Connection) *wsLink {
+	var c *wsLink
+	c = &wsLink{
+		WSLink: ws.NewWSLink(addr, conn, pingPeriod, func(msg *msgjson.Message) *msgjson.Error {
+			return handleMessage(c, msg)
+		}),
 		respHandlers: make(map[uint64]*responseHandler),
 	}
-}
-
-// Send sends the passed Message to the websocket client. If the client's
-// channel if blocking (outBufferSize pending messages), the client is
-// disconnected.
-func (c *wsLink) Send(msg *msgjson.Message) error {
-	if c.off() {
-		return ErrClientDisconnected
-	}
-	b, err := json.Marshal(msg)
-	if err != nil {
-		return err
-	}
-	select {
-	case c.outChan <- b:
-	case <-c.quit:
-		return ErrClientDisconnected
-	}
-	return nil
-}
-
-// Request sends the message to the client and tracks the response handler.
-func (c *wsLink) Request(msg *msgjson.Message, f func(conn Link, msg *msgjson.Message)) error {
-	c.logReq(msg.ID, f)
-	return c.Send(msg)
+	return c
```

Comment:
Somewhat verbose here.  How about just `return &wsLink{...`?

Reply:
I'm using `c` as an argument to `handleMessage`, which is a little odd, I admit.